### DISCUSSION TITLE
html template reloading

### DISCUSSION THREAD

**Discussion by aaronsgithub** [nickjj/docker-django-example#32] [Q&A]
Great work on this Nick.

I just want to check if my understanding of html template reloading is correct.

With the normal django development server, setting the `DEBUG` flag in the `TEMPLATES` dict in `settings.py` allows changes to templates to be loaded in the browser when the browser is refreshed.

However with gunicorn as the server, this does not work.
```
TEMPLATES = [
    {
        "BACKEND": "django.template.backends.django.DjangoTemplates",
        "DIRS": [os.path.join(BASE_DIR, "templates")],
        "APP_DIRS": True,
        "OPTIONS": {
            "context_processors": [
                "django.template.context_processors.debug",
                "django.template.context_processors.request",
                "django.contrib.auth.context_processors.auth",
                "django.contrib.messages.context_processors.messages",
            ],
            "debug": DEBUG,    # <- this has no effect
        },
    },

```
Is this the expected behaviour or am I doing something wrong?

The only way I've managed to get templates to reload is by adding the following to `gunicorn.py` (as per the docs [here](https://docs.gunicorn.org/en/stable/settings.html#reload-extra-files)):
```
from django.conf import settings
...

reload_extra_files = [str(settings.BASE_DIR / "<app-name>" / "templates" / "<app-name>" / "<template-name>.html")]
```
And so, to get the browser to load changes to any template file, I would either have to explicitly list those templates in `reload_extra_files` or I could probably populate this list programmatically.

There is a [pull request](https://github.com/benoitc/gunicorn/pull/1846) for gunicorn that would make this simpler, but unfortunately it has been sitting unmerged for 4 years.

Is there a better way?

**Comment by nickjj** [comment]
Hi,

We pass our app into gunicorn and then gunicorn is configured with:
https://github.com/nickjj/docker-django-example/blob/37cd9f8eb9093f8d2d4660370b75821cd6205fba/src/config/gunicorn.py#L14

That should handle all reloading by itself. It does for this example app which has a custom `pages` app with a `templates/pages/home.html` file. When that file changes, you can reload your browser and see it without customizing anything else.

**Comment by aaronsgithub** [comment]
Thanks for the quick reply.

I did see that setting, however for me changes to the templates to not appear in the browser after reload.

I've just cloned the repository and checked this in the pages app.

Two things suggested to me that we shouldn't expect the reload to work:

1.  The gunicorn docs say to use the [reload_extra_files](https://docs.gunicorn.org/en/stable/settings.html?highlight=reload#reload-extra-files) "to also watch and reload on additional files (e.g., templates, configurations, specifications, etc.)" which would suggest the reload setting alone is not enough for template files.

2. The [pull request](https://github.com/benoitc/gunicorn/pull/1846) on the gunicorn repo with users suggesting the same as 1. 

But again I'm not certain if this is the case or if the templates should reload on change and there is something wrong with my setup.

Have you been able to confirm the template reloading works recently?

---
Edit

So I'm developing on Ubuntu 22.04 on WSL2 on Windows 11. 

If I try to edit a template from WSL then as I wrote above the templates do not reload.

However, if I edit inside a [devcontainer](https://code.visualstudio.com/docs/devcontainers/containers) i.e. insider the docker container itself, the change is detected and gunicorn reloads so I guess in my setup, gunicorn is not seeing the file system changes originating from the WSL bind mount.

In any case, I like to develop inside the devcontainer so its not an issue for me.

**Reply by nickjj** [reply]
I'm using Ubuntu 22.04 and WSL 2 too, but on Windows 10.

Are your files in your WSL 2 file system? I think it might not work if your files are in a Windows drive that you're mounting into WSL 2. In my case my files are in WSL 2's file system.

**Reply by aaronsgithub** [reply]
The files are in the WSL2 file system.

Looking at [one of the old issue threads](https://github.com/nickjj/docker-django-example/issues/30#issuecomment-1307628781) which looks similar, it could be that I'm using neovim and running into the same issue which that affected user identified.

**Reply by nickjj** [reply]
It's possible but that would potentially only apply to CSS files due to a bug in Tailwind and you mentioned Django templates were causing an issue with not being reloaded. It's worth a shot delving down that rabbit hole tho or at least seeing if it happens with other editors as a quick test.

I use Vim too btw. My vimrc is at: https://github.com/nickjj/dotfiles/blob/master/.vimrc
